### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by maov19**
### Implementations
1. Installed CanCanCan.

3. Added a role column to the users table. 

4. A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). 
> Used CanCanCan for this authorization.

5. A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). 
> Used CanCanCan for this authorization.

Thanks! 🐦

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -18,4 +18,8 @@ def recent_posts(limit = 3)
   def update_posts_counter
     update(posts_counter: posts.count)
   end
+
+  def admin?
+    role == 'admin'
+  end
```
Comment (anagudelogu) [inline app/models/user.rb:24]: Kudos for implementing this method! 🚀

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new
+    can :destroy, Post, author_id: user.id
+    can :destroy, Comment, author_id: user.id
+    can :destroy, Post, author_id: user.id if user.admin?
+    can :destroy, Comment, author_id: user.id if user.admin?
```
Comment (anagudelogu) [inline app/models/ability.rb]: - [ ] According to CanCanCan [docs](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md):
> By default, CanCanCan assumes no permissions: no one can do any action on any object.

This means that at the moment, your app only allows `:destroy` actions. Also, naming the abilities one by one seems kind of repetitive, remember to keep the code DRY. Consider doing something like the following:

```ruby
         if user.admin? 
           can :manage, :all
         else
           can :manage, Post, author: user
           can :manage, Comment, author: user
           can :read, :all
         end
```

Take this code as a reference to implement yours! 🤓 🔧 The `manage` method refers to performing all CRUD actions.

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -25,6 +25,12 @@ def create
     end
   end
 
+  def destroy
+    @post = Post.find(params[:id])
+    @post.destroy
+    redirect_to user_path(@post.author), notice: 'Post deleted successfully'
+  end
+
```
Comment (anagudelogu) [inline app/controllers/posts_controller.rb:35]: - [ ] I see that you implemented the `destroy` method in your controllers, however, you haven't added the authorization to your actions! 😱 Please consider taking a look at this article from [CanCanCan official docs](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/controller_helpers.md#authorize_resource-load_resource-load_and_authorize_resource) on loading and authorizing resources and implement this! 🤓 🔧

File: app/controllers/comments_controller.rb [hunk 4]
```diff
@@ -15,6 +15,12 @@ def create
     end
   end
 
+  def destroy
+    @comment = Comment.find(params[:comment_id])
+    @comment.destroy
+    redirect_to user_post_path(@comment.author), notice: 'Comment was deleted succesfully'
+  end
+
```
Comment (anagudelogu) [inline app/controllers/comments_controller.rb:25]: - [ ] The same happens on this controller, please consider loading and authorizing the resource! 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS:  CHANGES REQUIRED  ♻️ 🚧 

Hi @maov19 @edubew, 
I'm Andres.


While you have done a great job working on this milestone 💻 , there are some issues that you still need to work on to go to the next project. ♻️ 

You are almost there, keep up the good work! 🚀 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Great UI design. 💻 
✅ Descriptive PR title, summary and commit messages. 💬 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 1 above)
Comment (maov19) [reply app/models/user.rb:24]: thanks! 💪

File: app/models/ability.rb (hunk 2 above)
Comment (maov19) [reply app/models/ability.rb]: fixed, thanks for the guidance 💯

File: app/controllers/posts_controller.rb (hunk 3 above)
Comment (maov19) [reply app/controllers/posts_controller.rb:35]: added for both, thanks for the observation!

File: app/controllers/comments_controller.rb (hunk 4 above)
Comment (maov19) [reply app/controllers/comments_controller.rb:25]: added for both, thanks for the observation!

**Review by t-yanick — APPROVED** [review]
# STATUS: MILESTONE APPROVED :1st_place_medal: 

Hi @maov19 , 

### Great job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  ALL changes correctly implemented
:heavy_check_mark: Excellent communication with reviewer

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**